File: ProjectEuler/Solver4.cs
using System;

namespace ProjectEuler
{
    class Solver4 : ISolver
    {
        public int Solve()
        {
            return Calculate(999);
        }

        static int Calculate(int i)
        {
            return (i > 100) ? Math.Max(Calculate(i, i - 1), Calculate(i - 1)) : 0;
        }

        static int Calculate(int i, int j)
        {
            var product = i * j;
            return IsPalindrome(product) && i != j ? product : j > 100 ? Calculate(i, j - 1) : 0;
        }

        static bool IsPalindrome(int number)
        {
            var str = number.ToString();
            var charArray = str.ToCharArray();
            Array.Reverse(charArray);
            return str == new string(charArray);
        }
    }
}

File: C/Solver4.c
#include <stdio.h>
#include <string.h>
#include "Common.h"

BOOL isPalindrome(int number)
{
	char str[] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
	char *start, *end;

	sprintf_s(str, 10, "%i", number);
	start = str;
    end = str + strlen(str) - 1;

	while(start < end)
	{
		if (*(start++) != *(end--))
			return FALSE;
	}

    return TRUE;
}

int calculateAndCheckPalindrome(int i, int j)
{
    int product = i * j;
    return i != j && isPalindrome(product) ? product : j > 100 ? calculateAndCheckPalindrome(i, j - 1) : 0;
}

int calculate4(int i)
{
    return (i > 100) ?  max(calculateAndCheckPalindrome(i, i - 1), calculate4(i - 1)) : 0;
}

int Solver4()
{
    return calculate4(999);
}